test/surgex/changeset/changeset_test.exs

```diff
@@ -2,32 +2,97 @@ defmodule Surgex.ChangesetTest do
   use ExUnit.Case
   alias Surgex.Changeset
 
+  import Ecto.Changeset, only: [change: 2, add_error: 3, add_error: 4]
+
+  defmodule TestSchema do
+    use Ecto.Schema
+    import Ecto.Changeset
+
+    embedded_schema do
+      field(:name, :string)
+
+      embeds_one :address, TestAddress do
+        field(:city, :string)
+        field(:street, :string)
+
+        embeds_one :country, TestCountry do
+          field(:code, :string)
+        end
+      end
+    end
+
+    def changeset(schema, params) do
+      schema
+      |> cast(params, [:name])
+      |> cast_embed(:address, with: &address_changeset/2)
+    end
+
+    defp address_changeset(schema, params) do
+      schema
+      |> cast(params, [:city, :street])
+      |> validate_required([:city])
+      |> cast_embed(:country, with: &country_changeset/2)
+    end
+
+    defp country_changeset(schema, params) do
+      schema
+      |> cast(params, [:code])
+      |> validate_inclusion(:code, ["us", "ca", "pl"])
+    end
+  end
+
   test "changeset with all possible errors" do
-    assert Changeset.build_errors_document(%{
-             errors: [
-               taken_field: {"has already been taken", nil},
-               required_field: {nil, [validation: :required]},
-               invalid_field: {nil, nil},
-               invalid_type_field: {nil, [validation: :type]}
-             ]
-           }) == %Jabbax.Document{
+    changeset =
+      %TestSchema{}
+      |> change(%{})
+      |> add_error(:taken_field, "has already been taken")
+      |> add_error(:required_field, "", validation: :required)
+      |> add_error(:invalid_type_field, "", validation: :type)
+      |> add_error(:invalid_field, "")
+
+    assert Changeset.build_errors_document(changeset) == %Jabbax.Document{
              errors: [
                %Jabbax.Document.Error{
-                 code: "taken",
-                 source: %Jabbax.Document.ErrorSource{pointer: "/data/attributes/taken_field"}
+                 code: "invalid",
```

Comment:
Shouldn't we keep the order of errors? I don't think it makes a huge difference, but some API/controller tests could fail if we change the order.

Reply:
I think `traverse_errors` forces its own order of errors. No matter in what order I use `add_error`, it does not change anything (even though it changes order in `errors` list in the changeset). I don't see any easy way to replay original order. It's not simply reversed.

New version returns: `invalid, invalid_type, required, take`
Old version was: `taken, required, invalid, invalid_type`

Reply:
Ok, let's build a pre-release version and see how it works.